js/autocomplete.js

```diff
@@ -13,26 +13,27 @@
           if (1 > --count) {
             $form
 
-              // Unbind submit listener so it doesn't preventDefault()
-              .unbind('submit')
+              // Unbind preSubmit listeners so they don't re-fire on submit
+              .off('preSubmit')
 
               .submit();
           }
         }
 
         // Fire preSubmit events, then let done() call the actual form submit
         $form.submit(function (event) {
-          event.preventDefault();
 
           // Trigger all "preSubmit" listeners
           $form.trigger('preSubmit');
 
-          // If no preSubmits fired, then unbind this submit handler, and fire a
-          // new submit event
+          // If no preSubmits fired, then continue with form submit as usual
           if (count == 0) {
-            $form.off('submit');
-            $form.submit();
+            return true;  // Go ahead and submit form
           }
```

Comment:
Could it be a race condition while the preSubmit listeners increment the count? Should we do this synchronously? It's hard to tell, but looking at the stable/2.5.x version, it seems like we only change the main form submit after interactions are made over the autocomplete fields. I really like that you called it `preSubmit`, because reading the stable/2.5.x version is quite a puzzle.

Reply:
Yeah, the logic is confusing.  The basic code path is this:

1. When a new related resource (e.g. a subject access point) is added to an autocomplete dialog (a new name is typed into the field by a user), add an iframe which contains a form to add the new related resource (e.g. subject),
2. On form submit (main page), fire the preSubmit event, which submits each iframe form and increments the counter,
3. When an iframe returns a response with the new related resource (e.g. new subject) copy the URI for the new resource into the main form and decrements the counter,
4. If the counter is less than 1 (e.g. all iframes responses are complete), then fire the main form submission,

If you are only linking existing resources, than there should be no iframes, no preSubmit listeners, and the form submit event should .

But, this is all complicated with the authority record relations, because the dialog.js code attaches it's own "onSubmit" listeners, and does it's own DOM manipulation before allowing the form submit (HTTP POST).

I don't know if race conditions are possible when the preSubmit listeners increment the count, but I haven't had problems so far with the form submit not happening, or with the submit firing before all of the iframes responses return - i.e. it seems to be working. :)

When you say "should we do this synchonously" do you mean firing the iframe submits in serial, or are suggesting using promises (w/ async/await?) to serialize the callback behaviour?

Reply:
Yeah, that makes sense, thanks @djjuhasz! 

> When you say "should we do this synchronously" do you mean firing the iframe submits in serial, or are you suggesting using promises (w/ async/await?) to serialize the callback behaviour?

It doesn't have to be in serial, you could submit all at once. I think the issue would be to use the event trigger/listener as that happens async. If you could submit the iframes directly in here, without using the event, that will avoid the race condition. You could probably remove the use of the count going this way, so I'm not sure if this is the way to go. I think the biggest change between 2.5 and now is how the main form submit is modified.

Reply:
Okay, I've made the iframe submit() callers synchronous.